api/v1alpha4/awsloadbalancerspec.go

```diff
@@ -0,0 +1,61 @@
+/*
+Copyright 2021 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package v1alpha4
+
+import (
+	"fmt"
+	"strings"
+
+	"github.com/pkg/errors"
+	"sigs.k8s.io/cluster-api-provider-aws/pkg/hash"
+)
+
+// GenerateELBName generates a formatted ELB name via either
+// concatenating the cluster name to the "-apiserver" suffix
+// or computing a hash for clusters with names above 32 characters.
+func GenerateELBName(clusterName string) (string, error) {
```

Comment:
Do we think the generation of the ELB name is likely to change between API versions?

Reply:
I'm wondering if we introduce something like `pkg/generate` as there are a few places that we generate a name of a thing and then if its too long do some form of hashing.

Reply:
> Do we think the generation of the ELB name is likely to change between API versions?

Good question.

If the generation of the ELB name changes, the AWSCluster controller _must_ use the existing ELB name, because the existing ELB must not be replaced. I think that implies one of the following:

1. The generated name must be stored in the AWSCluster object. _Since it is not possible for a defaulting webhook to generate and store the generated name (I explain why this is in the PR description), I think the only option is for AWSCluster controller to store the generated name. To make this possible, the validating webhook would allow an empty string name initially, but make the field immutable after._

2. The controller should discover the existing ELB something _other_ than the name, e.g., a tag. _This allows the generated name to change, but for the controller to use an existing ELB. However, an ELB name is a primary key, and therefore unique; tags are not, and this option creates a new failure mode where more the controller finds more than one ELB._